src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1089,16 +1103,94 @@ private static void AddWrapperToReferenceTrackerHandleCache(NativeObjectWrapper
             }
         }
 
-        private void RemoveRCWFromCache(IntPtr comPointer, GCHandle expectedValue)
+        private sealed class RcwCache
```

Comment:
I agree that moving the RCW cache logic into this separate type helps with maintainability, but I'm wondering if we could make this a `struct` (and make the `_rcwCache` field not readonly) to avoid an extra indirection every time we need to use it, which is pretty often? We don't really need this to be a class anyway, it's only ever used internally in this file, and it doesn't have a finalizer or anything. Thoughts? 🙂

Reply:
It makes little sense to turn this into a `struct`. Every method here takes a lock. The indirection is noise in most cases and the lock means it won't even be measurable in practice.